packages/flutter/lib/src/widget_preview/widget_preview.dart

```diff
@@ -0,0 +1,302 @@
+  }
+}
+
+/// Wrapper applying a custom render object to force constraints on
+/// unconstrained widgets.
+@visibleForTesting
+class WidgetPreviewWrapper extends SingleChildRenderObjectWidget {
+  @visibleForTesting
+  // ignore: public_member_api_docs
+  const WidgetPreviewWrapper({
+    super.key,
+    super.child,
+    required this.previewerConstraints,
+  });
+
+  /// The ratio of the max height provided by a parent
+  /// [WidgetPreviewerWindowConstraints] that an unconstrained child should
+  /// be allowed to occupy.
+  @visibleForTesting
+  static const double unconstrainedChildScalingRatio = 0.5;
+
+  /// The size of the previewer render surface.
+  final BoxConstraints previewerConstraints;
+
+  @override
+  RenderObject createRenderObject(BuildContext context) {
+    return WidgetPreviewWrapperBox(
+      previewerConstraints: previewerConstraints,
+      child: null,
+    );
+  }
+
+  @override
+  void updateRenderObject(
+    BuildContext context,
+    WidgetPreviewWrapperBox renderObject,
+  ) {
+    renderObject.setPreviewerConstraints(previewerConstraints);
+  }
+}
+
+/// Custom render box that forces constraints onto unconstrained widgets.
+@visibleForTesting
+class WidgetPreviewWrapperBox extends RenderShiftedBox {
+  // ignore: public_member_api_docs
+  WidgetPreviewWrapperBox({
+    required RenderBox? child,
+    required BoxConstraints previewerConstraints,
+  })  : _previewerConstraints = previewerConstraints,
+        super(child);
+
+  BoxConstraints _constraintOverride = const BoxConstraints();
+  BoxConstraints _previewerConstraints;
+
+  /// Updates the constraints for the child based on changes to the constraints
+  /// provided by the parent [WidgetPreviewerWindowConstraints].
+  void setPreviewerConstraints(BoxConstraints previewerConstraints) {
+    if (_previewerConstraints == previewerConstraints) {
+      return;
+    }
+    _previewerConstraints = previewerConstraints;
+    markNeedsLayout();
+  }
+
+  @override
+  void layout(
+    Constraints constraints, {
+    bool parentUsesSize = false,
+  }) {
+    if (child != null && constraints is BoxConstraints) {
+      double minInstrinsicHeight;
+      try {
+        minInstrinsicHeight = child!.getMinIntrinsicHeight(
+          constraints.maxWidth,
+        );
+      } on Object {
+        minInstrinsicHeight = 0.0;
+      }
+      // Determine if the previewed widget is vertically constrained. If the
+      // widget has a minimum intrinsic height of zero given the widget's max
+      // width, it has an unconstrained height and will cause an overflow in
+      // the previewer. In this case, apply finite constraints (e.g., the
+      // constraints for the root of the previewer). Otherwise, use the
+      // widget's actual constraints.
+      _constraintOverride = minInstrinsicHeight == 0
+          ? _previewerConstraints
+          : const BoxConstraints();
+    }
+    super.layout(
+      constraints,
+      parentUsesSize: parentUsesSize,
+    );
+  }
```

Comment:
Since `layout` is not supposed to be overwritten: Could this logic just live in `performLayout`? From my understanding it should only matter that the child gets the updated constraints?